feat(probot): github app adds empty commit if PR has 1 single commit

oscard0m opened the pull request:
## 📝 Summary
<!--- Provide a general summary of your changes, and if you feel that the commit comments are not descriptive enough, give more detail of your changes -->
GitHub App adds empty commit if PR has 1 single commit.

## ⛱ Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
When using "Squash and merge" on a PR with only one commit, GitHub will suggest using that commit message instead of the PR title for the merge commit, and it's easy to commit this by mistake. This can be specially problematic when working with [Semantic Releases](https://github.com/zeke/semantic-pull-requests/issues/106).

The solution we are proposing is to create an empty commit so a PR is not single-commit anymore and, for scenarios like the one for Semantic Release, only PR title will be taken in consideration.

Fix #1

## 🏗 TODO
- [x] Tests
- [x] Deployment

## 📊 How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
* Running probot app in local
* PR examples where the app is working: 
https://github.com/oscard0m/sandbox/pull/35/
https://github.com/oscard0m/sandbox/pull/36

## 📸 Screenshots:
#### PR with empty commit
![image](https://user-images.githubusercontent.com/2574275/116869133-70fcc500-ac10-11eb-8a15-5bdbad026592.png)

#### Local logs
![image](https://user-images.githubusercontent.com/2574275/116869303-c6d16d00-ac10-11eb-875b-73cfe3d70421.png)


## 📚 References:
<!-- Any interesting external link to documentation, article, tweet which can add value to the PR -->
* https://github.com/octokit/rest.js/discussions/2018#discussioncomment-399606
* https://github.com/zeke/semantic-pull-requests/issues/106

oscard0m committed:
feat(probot): github app adds empty commit if PR has 1 single commit

oscard0m committed:
build(npm): build package.json and install deps

oscard0m committed:
build(typescript): add tsconfig.json to project

oscard0m committed:
docs(code-of-conduct): add code of conduct to project

oscard0m committed:
build(gitignore): add gitignore to project

oscard0m commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+.env
```
Not sure if this will be necessary

gr2m replied:
it's good for local development

oscard0m commented on app.yml:
```diff
@@ -0,0 +1,7 @@
+
+default_permissions:
+  pull_requests: read
+  metadata: read
```
Not sure if this will be necessary. When we run the Github App we will see.

gr2m replied:
it's required by default. I think we get it even if we don't set it here

oscard0m commented on tsconfig.json:
```diff
@@ -0,0 +1,6 @@
+{
+	"extends": "@tsconfig/node14/tsconfig.json",
```
`node v14` is ok for you?

gr2m replied:
sure

oscard0m commented on src/app.ts:
```diff
@@ -0,0 +1,90 @@
+        owner,
+        repo,
+        message: 'empty commit',
+        tree: '', //TODO To check how to get the SHA of the tree object this commit points to
```
🏗️ To check how to get the SHA of the tree object this commit points to

oscard0m replied:
Will we need all the logic in this discussion necessary for getting the tree? https://github.com/octokit/rest.js/discussions/1938#discussioncomment-132896

oscard0m replied:
Inspired by this issue comment: https://github.com/renovatebot/renovate/issues/4408#issuecomment-664012687

oscard0m commented on src/app.ts:
```diff
@@ -0,0 +1,90 @@
+module.exports = (app: Probot): void => {
+  app.log("Yay! The app was loaded!");
+
+  //TODO pull_request.opened should be an event too?
```
Would it be possible a user opens a PR, gets reviewed (or no reviewer is required) and it tries to merge the PR. In this case, `pull_request.edited` event would not trigger right?

gr2m replied:
right

oscard0m replied:
I just assume the app needs to run on:
- `pull_request.opened`: First time a PR is opened
- `pull_request.synchronize`: Everytime the head changes.
---
The """problem""" is that it will run twice everytime an empty commit is done.

gr2m replied:
you mean it will get triggered by itself? That's okay, we can check the author of the commit is the app itself, and ignore it in that case

oscard0m commented on src/app.ts:
```diff
@@ -0,0 +1,90 @@
+  const { data: { sha } } = await context.octokit.git.createCommit({
+        owner,
+        repo,
+        message: 'empty commit',
```
Should we consider this commit message configurable (via `squash-commit-app.yml`) for example? If you agree I will open an issue where we can discuss this further.

gr2m replied:
let worry about configuration later. For now, I'd set a more descriptive commit message, something like this
```suggestion
        message: `empty commit to preset the squash & merge commit subject from the pull request title
        
Created by https://github.com/gr2m/squash-commit-app`,
```

gr2m commented:
I'll have a look after the GitHub event tomorrow: https://www.meetup.com/GitHub-Virtual-Meetup/events/273139225/

G-Rath commented on src/api/github/webhooks/index.ts:
```diff
@@ -0,0 +1,10 @@
+const { createNodeMiddleware, createProbot } = require("probot");
```
Any reason why we're using CommonJS `require/module.exports` here when we're using `import/export` elsewhere? (and am using TS, so generally should always just be able to use `import/export`)

oscard0m replied:
> Any reason why we're using CommonJS `require/module.exports` here when we're using `import/export` elsewhere? (and am using TS, so generally should always just be able to use `import/export`)

I just removed the code but yep, you are right.

G-Rath commented on src/app.ts:
```diff
@@ -0,0 +1,90 @@
+
+  //TODO pull_request.opened should be an event too?
+  app.on("pull_request.edited", async (context) => {
+    const { pullRequest } = context
+
+    const pullRequestInfo = pullRequest();
```
```suggestion
    const pullRequestInfo = context.pullRequest();
```

nit: I don't think it's worth a variable given it's just being used the once.

G-Rath commented on src/app.ts:
```diff
@@ -0,0 +1,90 @@
+      owner, repo, pull_number
+    });
+  
+  return { prNumCommits: commits, prHeadSha: head.sha, prBranchName: head.ref }
```
nit: I don't think we need to prefix these properties with `pr` since the function is called `getPRInfo` so it should be expected that the info these properties have are for the PR.

This'll let you use shorthand properties in later code :)

gr2m commented on src/api/github/webhooks/index.ts:
```diff
@@ -0,0 +1,10 @@
+  },
+});
+
+module.exports = createNodeMiddleware(app, { probot });
```
why the `src/` prefix? I think Vercel supports TypeScript out of the box, and the webhook handler function needs to be in the `api/` folder.

oscard0m replied:
I still have pending the deploy part. For the moment I just removed this file from the draft PR. And, once validated, I will focus on learning how vercel deployment for probot works :)

G-Rath commented on src/app.ts:
```diff
@@ -0,0 +1,90 @@
+
+      return pushCommitToBranch(context, {
+        ...pullRequestInfo,
+        branchName: prBranchName,
```
nit: You could shorthand this property if you adjusted the return value of `getPRInfo` (per my comment further up)

gr2m commented on src/app.ts:
```diff
@@ -0,0 +1,90 @@
+      prBranchName
+    } = await getPRInfo(context, pullRequestInfo)
+
+    if (prNumCommits === 1) {
+      const emptyCommitSha = await createEmptyCommit(context, {
+        ...pullRequestInfo,
+        prHeadSha
+      })
+
+      return pushCommitToBranch(context, {
+        ...pullRequestInfo,
+        branchName: prBranchName,
+        commitSha: emptyCommitSha
+      })
+    }
+
+    return 'not a single commit PR';
```
I'd return early instead, it makes the code more readable

```suggestion
    if (prNumCommits > 1) return 'not a single commit PR'
    
    const emptyCommitSha = await createEmptyCommit(context, {
      ...pullRequestInfo,
      prHeadSha
    })

    return pushCommitToBranch(context, {
      ...pullRequestInfo,
      branchName: prBranchName,
      commitSha: emptyCommitSha
    })
  }
```

gr2m reviewed (commented):
I didn't  run it myself yet either, I'm sure there are bugs in there that will be easier to debug once you run it :)

oscard0m committed:
build(deps): add typescript as devDependency

oscard0m committed:
fix(app): apply proper octokit git usage

oscard0m committed:
chore(commit-message): improve empty commit message

oscard0m committed:
refactor(app): remove unused code

oscard0m committed:
refactor(app): no else if return before

oscard0m committed:
chore(app-logging): improve logging message

oscard0m committed:
ref(app): do not rename pr properties in function

oscard0m commented:
@gr2m Still pending deployment part (which service should I use?) and tests but I would like you to validate the app first 🙏🏽

gr2m commented:
> which service should I use?)

Let's deploy to vercel, it might be the easiest. Eventually we can experiment with other deployments to see if there is any difference in speed, as the time it takes the app to add the empty commit is critical.

gr2m commented on package.json:
```diff
@@ -0,0 +1,27 @@
+  "name": "squash-commit-app",
+  "version": "1.0.0",
+  "private": true,
+  "description": "Probot & Google Cloud Functions example",
```
```suggestion
  "description": "GitHub App which adds an empty commit to an open pull request with a single commit when the title was changed",
```

gr2m commented on src/app.ts:
```diff
@@ -0,0 +1,130 @@
+
+        if (commits > 1) {
+          context.log.info("not a single commit PR");
+          return "not a single commit PR";
```
The string return has no function, I'd remove it

```suggestion
          return;
```

gr2m commented on src/app.ts:
```diff
@@ -0,0 +1,130 @@
+          pullRequestInfo.repo
+        );
+
+        const empty = await createEmptyCommit(
+          context,
+          commit,
+          pullRequestInfo.owner,
+          pullRequestInfo.repo
+        );
+
+        return updateRef(
+          context,
+          empty.sha,
+          ref,
+          pullRequestInfo.owner,
+          pullRequestInfo.repo
+        );
```
I think updating the ref should be part of the `createEmptyCommit` function, but we can change that later.

Eventually this would make a great octokit plugin :)

oscard0m replied:
> I think updating the ref should be part of the `createEmptyCommit` function, but we can change that later.

Good point

> Eventually this would make a great octokit plugin :)

I was thinking the same when I was writing these lines 😇 . Created card https://github.com/users/oscard0m/projects/1#card-60464075

gr2m reviewed (commented):
code looks good to me! Did you test it locally?

oscard0m commented:
> code looks good to me! Did you test it locally?

Yep. I tested it locally :)

oscard0m committed:
docs(package.json): update description of the package

oscard0m committed:
build(deps): update package-lock.json

oscard0m committed:
chore(app): log info when running getPRInfo and getCommit steps

oscard0m committed:
refactor(app): remove unnecessary string return

oscard0m committed:
refactor(app): move updateRef logic inside createEmptyCommit method

oscard0m committed:
revert(refactor-app): remove unused code"

This reverts commit 653f0e31c40129e343f45c6980eee88eaa44492c.

oscard0m commented:
Hey @gr2m,

I just did a [fork](https://github.com/oscard0m/squash-commit-app) to restore `/api/github/webhooks`, install Vercel's GitHub App and deploy it. I think I'm missing something. Even the app was supposedly deployed properly:

<summary>Build Logs</summary>
<details>



```
10:46:01.050  	Cloning github.com/oscard0m/squash-commit-app (Branch: main, Commit: db726bb)
10:46:01.519  	Cloning completed: 469.221ms
10:46:01.545  	Analyzing source code...
10:46:02.896  	Installing build runtime...
10:46:05.787  	Build runtime installed: 2.890s
10:46:09.630  	Build cache not provided
10:46:11.035  	Detected package.json
10:46:11.035  	Installing dependencies...
10:46:11.637  	npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!
10:46:18.591  	added 353 packages from 308 contributors in 6.998s
10:46:18.718  	23 packages are looking for funding
10:46:18.718  	  run `npm fund` for details
10:46:19.274  	> squash-commit-app@1.0.0 build /vercel/path0
10:46:19.275  	> tsc
10:46:22.359  	Uploading build outputs...
10:46:22.412  	Deploying build outputs...
10:46:28.954  	Build completed. Populating build cache...
10:46:32.899  	Uploading build cache [13.68 MB]...
10:46:34.376  	Build cache uploaded: 1.477s
10:46:34.895  	Done with "package.json"
```


</details>

And the files are properly compiled and available: https://squash-commit-app.vercel.app/api/github/webhooks
When doing a `POST` to `https://squash-commit-app.vercel.app/api/github/webhooks` I get a `405` response (with Postman or as Webhook URL in my GitHub App).

Is there any `.vercel/` or `vercel.json` I could be missing? I did a quick research and I did not find a solution. These are my `Settings` in case it can give you any idea:

![image](https://user-images.githubusercontent.com/2574275/117122063-b991bb00-ad95-11eb-93ed-e3140dc306a3.png)

Once we find the issue I will update https://github.com/probot/example-vercel with the solution ;)

oscard0m commented:
> Hey @gr2m,
> 
> ....

Mistery solved. I re-read the comments you left to me and all the solutions were there:
- TS is supported out of the box
- `api/` folder [MUST](https://github.com/vercel/vercel/discussions/4592#discussioncomment-26597) be outside `src/`
- Added missing environment variables to Vercel (to avoid `400 signature does not match event payload and secret` error)

oscard0m committed:
fix(vercel): apply tweaks necessary to Vercel deployment

* Initial version GitHub app (#1)

* feat(probot): github app adds empty commit if PR has 1 single commit

* build(npm): build package.json and install deps

* build(typescript): add tsconfig.json to project

* docs(code-of-conduct): add code of conduct to project

* build(gitignore): add gitignore to project

* build(deps): add typescript as devDependency

* fix(app): apply proper octokit git usage

* chore(commit-message): improve empty commit message

* refactor(app): remove unused code

* refactor(app): no else if return before

* chore(app-logging): improve logging message

* ref(app): do not rename pr properties in function

* docs(package.json): update description of the package

* build(deps): update package-lock.json

* chore(app): log info when running getPRInfo and getCommit steps

* refactor(app): remove unnecessary string return

* refactor(app): move updateRef logic inside createEmptyCommit method

* build(deploy): restore entrypoint file for vercel deployment (#2)

* fix(deploy): vercel entry points must be at root level as .js

* fix(vercel): use TS properly for deployment with Vercel + TS

https://vercel.com/docs/serverless-functions/supported-languages#using-typescript

* fix(vercel): add missing public/ folder

* fix(vercel): move api/ folder to root

* fix(api): update import to correspond to new file tree

oscard0m commented:
Now yes! Ready for review!

oscard0m committed:
refactor(api): remove unused file

oscard0m committed:
feat(app): act when a PR is reopened

oscard0m committed:
refactor(app): migrate to ESM exports

oscard0m committed:
test(app): add test for the app

gr2m committed:
trigger deploy to vercel

gr2m commented:
@oscard0m I've setup automated deployments with Vercel, but I need the credentials for https://github.com/apps/squash-commit-app. Can you dm me the app id and private key?

gr2m committed:
trigger deploy to vercel

gr2m committed:
trigger deploy to vercel

gr2m committed:
trigger deploy to vercel

gr2m commented:
I tested it on https://github.com/gr2m/sandbox/pull/190, the webhook went through, I see `[POST] /api/github/webhooks` in the logs, but no commit was created. Any idea?

gr2m commented on src/app.ts:
```diff
@@ -0,0 +1,133 @@
+    [
+      "pull_request.opened",
+      "pull_request.reopened",
+      "pull_request.synchronize",
```
I think we need to add `pull_request.edited` here, so that pull requests that existed before the app install can benefit from the app easily

```suggestion
      "pull_request.synchronize",
      "pull_request.edited",
```

oscard0m replied:
I was thinking in using `reopen` but sure, we can add `pulll_request.edited` if you want

gr2m commented:
Yep, it worked with a new PR:
https://github.com/gr2m/sandbox/pull/191

<img width="976" alt="image" src="https://user-images.githubusercontent.com/39992/117198498-f366bf80-ad9d-11eb-8258-1952af822e27.png">

oscard0m committed:
feat(app): listen to pull_request.edited event too

Co-authored-by: Gregor Martynus <39992+gr2m@users.noreply.github.com>

oscard0m committed:
test(app): add pull_request.edited event to tests

gr2m reviewed (approved):
all right let's go!
